### PR TITLE
Hotfix/multi gpu p2p bugfix

### PR DESCRIPTION
**Pull request by maddyscientist**
* Important bug fix for a latent race condition in the dslash: was triggered if in a given dimension, the MPI communication in a given direction completed before the local peer-to-peer communication  completed in the other direction in the same dimension.  This scenario only really applies to systems with first-class GPU RDMA support where MPI could complete in comparable time as peer-to-peer.
* Ensure that split kernel dslash halo kernels run in order T->X to prevent any possible race conditions if lower dimensions complete before high dimensions.  This ensures reproducibility of dslash computation. 
* Added option to environment variable `QUDA_ENABLE_P2P=-1`: if the process topology is length 2, then enabling this will only allow half of the network traffic to traverse peer-to-peer and force the other half to traverse MPI.  Useful to extrapolating performance at high node counts. 
* Improved reporting statistics for bi-directional bandwidth of dslash_test
* Targeted kernel and precision testing of blas_test, since this now respects `--test` and `--prec` flags to overwrite default behaviour

### CODE REVIEW

**Comment by mathiaswagner** [comment]
Wondering whether we should merge this before #650 to get this fix also in for the old MILC interface.

**Comment by maddyscientist** [comment]
@mathiaswagner just use 84a072f70e2a7b1e407e2557433fdee59862bd42 for old MILC: this has all the fixes without the feature/milc-site-optimize branch merged in.  Tag this and make that the pre0.9 release.

**Comment by mathiaswagner** [comment]
Re-merged develop as github complained about merge conflicts.

File: lib/comm_common.cpp [hunk 1]
```diff
@@ -142,6 +142,101 @@ void comm_destroy_topology(Topology *topo)
...
+	    cudaDeviceGetP2PAttribute(&accessRank[1], cudaDevP2PAttrPerformanceRank, neighbor_gpuid, gpuid);
+	  }
+#endif
+
+	  // enable P2P if we can access the peer or if peer is self
+	  if (canAccessPeer[0]*canAccessPeer[1] || gpuid == neighbor_gpuid) {
+	    peer2peer_enabled[dir][dim] = true;
+	    if (getVerbosity() > QUDA_SILENT) {
+	      printf("Peer-to-peer enabled for rank %d (gpu=%d) with neighbor %d (gpu=%d) dir=%d, dim=%d, performance rank = (%d, %d)\n",
+		     comm_rank(), gpuid, neighbor_rank, neighbor_gpuid, dir, dim, accessRank[0], accessRank[1]);
```
Comment (mathiaswagner) [inline lib/comm_common.cpp]: I guess accessRank is undefined here if gpuid ==neighbor_gpuid.

Comment (maddyscientist) [reply lib/comm_common.cpp]: No, it's initialized on line 199.

**Pull request merged**